File: app/lib/models/attachment.dart
import 'dart:typed_data';

/// A file hung off a task by a specific member.
///
/// Notes answer "what about this?"; attachments answer "here's the thing" — the
/// screenshot, the spec, the photo of the whiteboard. Authorship is a first-
/// class field rather than metadata because the board groups them by member,
/// one row each.
class Attachment {
  final String id;
  final String taskId;
  final String memberId;

  /// The name the person picked, not PocketBase's randomised stored name.
  final String name;

  /// Absolute download URL. Empty while [uploading].
  final String url;

  /// Server-rendered small version of [url], for grid previews. Falls back to
  /// [url] when the backend can't thumbnail this file.
  final String thumbUrl;

  /// Bytes in the file. 0 when unknown.
  final int size;

  final DateTime created;

  /// True between picking the file and the server confirming it.
  final bool uploading;

  /// Held in memory ONLY while [uploading], so an image previews instantly
  /// instead of after a round trip. Dropped once the real record arrives.
  final Uint8List? localBytes;

  const Attachment({
    required this.id,
    required this.taskId,
    required this.memberId,
    required this.name,
    required this.created,
    this.url = '',
    this.thumbUrl = '',
    this.size = 0,
    this.uploading = false,
    this.localBytes,
  });

  /// A placeholder for a file that's been picked but not yet stored. Image
  /// bytes ride along so the preview appears on tap; anything else would just
  /// be megabytes parked in memory for no visible gain.
  factory Attachment.pending({
    required String id,
    required String taskId,
    required String memberId,
    required String name,
    required Uint8List bytes,
  }) =>
      Attachment(
        id: id,
        taskId: taskId,
        memberId: memberId,
        name: name,
        created: DateTime.now(),
        size: bytes.length,
        uploading: true,
        localBytes: isImageName(name) ? bytes : null,
      );

  static const _imageExtensions = {
    'png',
    'jpg',
    'jpeg',
    'gif',
    'webp',
    'bmp',
    'heic',
    'heif',
    'avif',
  };

  /// Whether this should render as a picture rather than a file chip.
  bool get isImage => isImageName(name);

  static bool isImageName(String name) => _imageExtensions.contains(_extensionOf(name));

  /// Lowercase extension without the dot, or '' if the name has none.
  String get extension => _extensionOf(name);

  static String _extensionOf(String name) {
    final dot = name.lastIndexOf('.');
    if (dot < 0 || dot == name.length - 1) return '';
    return name.substring(dot + 1).toLowerCase();
  }

  /// "412 KB", "2.3 MB". Empty when the size is unknown.
  String get prettySize {
    if (size <= 0) return '';
    if (size < 1024) return '$size B';
    if (size < 1024 * 1024) return '${(size / 1024).round()} KB';
    return '${(size / (1024 * 1024)).toStringAsFixed(1)} MB';
  }
}

File: web/src/models/attachment.ts
import type { RecordModel } from "pocketbase";
import { readDate, readNum, readStr } from "./wire";

/* Port of app/lib/models/attachment.dart.
 *
 * Attachments are the EXISTING per-task feature: a screenshot pinned to a task,
 * stored in PocketBase's own file field, capped at 20MB, cascade-deleted with
 * the task. They are deliberately NOT the same thing as the shared drive in
 * models/fileRec.ts — see that file's header.
 *
 * The size/extension helpers below are kept because the file explorer reuses
 * them (plan.md §5); they are pure string arithmetic and belong to neither
 * store in particular. */

export interface Attachment {
  id: string;
  taskId: string;   // wire: "task"
  memberId: string; // wire: "member"
  /** The ORIGINAL filename. PocketBase randomises the name it stores, so this
   *  field is the only place the real one survives. */
  name: string;
  url: string;
  thumbUrl: string;
  size: number;
  created: Date;
  /** Client-only: a placeholder that has not reached the server yet. */
  uploading: boolean;
  /** Client-only: bytes held for an instant preview of an image mid-upload. */
  localPreview: string | null;
}

/* Built by hand rather than through `pb.files.getURL`, so the model layer stays
 * a pure function of the record and needs no client instance. This is the same
 * URL the SDK constructs.
 *
 * `thumbs` must list every size the app can ask for: PocketBase serves the
 * FULL-SIZE original for any thumb spec not declared in the migration, which on
 * a board full of phone photos is megabytes per tile. 240x240 is the only size
 * 1721700300_add_attachments.js declares. */
const fileUrl = (r: RecordModel, stored: string, thumb?: string): string => {
  if (!stored) return "";
  const base = `/api/files/${r.collectionId}/${r.id}/${encodeURIComponent(stored)}`;
  return thumb ? `${base}?thumb=${thumb}` : base;
};

export const toAttachment = (r: RecordModel): Attachment => {
  const stored = readStr(r.file);
  return {
    id: r.id,
    taskId: readStr(r.task),
    memberId: readStr(r.member),
    name: readStr(r.name),
    url: fileUrl(r, stored),
    thumbUrl: fileUrl(r, stored, "240x240"),
    size: readNum(r.size),
    created: readDate(r.created) ?? new Date(),
    uploading: false,
    localPreview: null,
  };
};

/* ---- shared filename helpers ------------------------------------------- */

const IMAGE_EXTENSIONS = new Set([
  "png", "jpg", "jpeg", "gif", "webp", "bmp", "heic", "heif", "avif",
]);

/** Lowercase, no dot, `''` when there is no extension or the dot is last. */
export function extensionOf(name: string): string {
  const i = name.lastIndexOf(".");
  if (i < 0 || i === name.length - 1) return "";
  return name.slice(i + 1).toLowerCase();
}

export const isImageName = (name: string): boolean =>
  IMAGE_EXTENSIONS.has(extensionOf(name));

/** "" / "812 B" / "43 KB" / "1.4 MB" — ported from Dart's prettySize. */
export function prettySize(bytes: number): string {
  if (bytes <= 0) return "";
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${Math.round(bytes / 1024)} KB`;
  const mb = bytes / (1024 * 1024);
  if (mb < 1024) return `${mb.toFixed(1)} MB`;
  return `${(mb / 1024).toFixed(1)} GB`;
}
